x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java

```diff
@@ -148,24 +96,143 @@ public LogicalPlan apply(LogicalPlan plan) {
+                    // It's an INLINEJOIN and all remaining attributes are groupings, which are already part of the IJ output (from the
+                    // left-hand side).
+                    if (aggregate.child() instanceof StubRelation stub) {
+                        var message = "Aggregate groups references ["
+                            + remaining
+                            + "] not in child's (StubRelation) output: ["
+                            + stub.outputSet()
+                            + "]";
+                        assert stub.outputSet().containsAll(Expressions.asAttributes(remaining)) : message;
+
+                        p = emptyLocalRelation(aggregate);
+                    } else {
+                        // There are no aggregates to compute, just output the groupings; these are already in the IJ output, so only
+                        // restrict the output to what remained.
+                        p = new Project(aggregate.source(), aggregate.child(), remaining);
+                    }
+                } else { // not an INLINEJOIN or there are actually aggregates to compute
+                    p = aggregate.with(aggregate.groupings(), remaining);
+                }
+            }
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(InlineJoin ij, AttributeSet.Builder used, Holder<Boolean> recheck) {
+        LogicalPlan p = ij;
+
+        used.addAll(ij.references());
+        var right = pruneColumns(ij.right(), used, true);
+        if (right.output().isEmpty()) {
+            p = ij.left();
+            recheck.set(true);
+        } else if (right != ij.right()) {
+            // if the right side has been updated, replace it
+            p = ij.replaceRight(right);
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(Eval eval, AttributeSet.Builder used, Holder<Boolean> recheck) {
+        LogicalPlan p = eval;
+
+        var remaining = pruneUnusedAndAddReferences(eval.fields(), used);
+        // no fields, no eval
+        if (remaining != null) {
+            if (remaining.isEmpty()) {
+                p = eval.child();
+                recheck.set(true);
+            } else {
+                p = new Eval(eval.source(), eval.child(), remaining);
+            }
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(Project project, AttributeSet.Builder used) {
+        LogicalPlan p = project;
+
+        var remaining = pruneUnusedAndAddReferences(project.projections(), used);
+        if (remaining != null) {
+            p = remaining.isEmpty() || remaining.stream().allMatch(FieldAttribute.class::isInstance)
+                ? emptyLocalRelation(project)
+                : new Project(project.source(), project.child(), remaining);
+        } else if (project.output().stream().allMatch(FieldAttribute.class::isInstance)) {
```

Comment:
Why are we interested in this aspect? (the fact that the projection is made of fields only)

Reply:
The explanation is the one given in the previous comment -- I've added a code comment too.